Source/CBLRouter+Changes.m

```diff
@@ -235,21 +257,50 @@ - (void) dbChanged: (NSNotification*)n {
         if (_changesMode == kLongPollFeed) {
             [changes addObject: rev];
         } else {
-            Log(@"CBL_Router: Sending continous change chunk");
             [self sendContinuousLine: [self changeDictForRev: rev]];
         }
+        _changesTimeoutSince = rev.sequence;
     }
 
     if (_changesMode == kLongPollFeed && changes.count > 0) {
-        Log(@"CBL_Router: Sending longpoll response");
-        [self sendResponseHeaders];
-        NSDictionary* body = [self responseBodyForChanges: changes since: 0];
-        _response.body = [CBL_Body bodyWithProperties: body];
-        [self sendResponseBodyAndFinish: YES];
+        [self sendLongpollResponseForChanges: changes since: 0];
+    } else {
+        if (changes.count > 0)
+            _changesTimeoutSince = ((CBL_Revision*)changes.lastObject).sequence;
+        [self startTimeout];
     }
 
     retainSelf = nil;
 }
 
 
+- (void)sendLongpollResponseForChanges: (NSArray*)changes since: (UInt64)since {
+    Log(@"CBL_Router: Sending longpoll response");
+    [self sendResponseHeaders];
+    NSDictionary* body = [self responseBodyForChanges: changes since: since];
+    _response.body = [CBL_Body bodyWithProperties: body];
+    [self sendResponseBodyAndFinish: YES];
+}
+
+
+- (void) startTimeout {
+    assert(_changesTimeout > 0);
+    [_changesTimeoutTimer invalidate];
+    _changesTimeoutTimer = [NSTimer scheduledTimerWithTimeInterval: _changesTimeout
```

Comment:
I think this creates a ref cycle because self retains the timer, but the timer's block retains self. So if the timeout isn't stopped, it creates a leak.
You could probably make `_changesTimeoutTimer` weak, because the runloop will hold onto it until the timer fires.